### ISSUE TITLE
Research: show GPU attached to a function

### ISSUE THREAD

**Issue by alexellis** [openfaas/faas#639]
## Description

Show a GPU attached to an OpenFaaS function in Kubernetes

## Background

We have several users using Python for data-science where GPU acceleration is available. From the investigation I've done so far we should be able to make a few minor changes to [faas-netes](https://github.com/openfaas/faas-netes/) and then be able to mount a GPU into a function.

## Tasks

* List compatible GPUs
* Write some code to mount a GPU
* Produce a short list of steps to document how to test the patches/PR
* Document any specific requirements / limitations

## Other notes

GKE has GPUs available pre-configured under Kubernetes - I think this would be the easiest way to test - https://thenewstack.io/getting-started-with-gpus-in-google-kubernetes-engine/

Otherwise you'll need an Nvidia GPU and the process for configuring your `kubelet` is not trivial

Documentation page from Kubernetes:

https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/

**Comment by LucasRoesler** [comment]
Per these docs https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/

There are two core changes

1. There is a new schedulable resource `nvidia.com/gpu`, this is a required change.
2. it is possible to have mixed types of resources, so they recommend using node lables and node selectors to ensure that your pod ends up on the node with the specific GPU you are looking for (this is probably optional and very advanced).


The simplest example of a pod using a GPU is provided as 

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cuda-vector-add
spec:
  restartPolicy: OnFailure
  containers:
    - name: cuda-vector-add
      # https://github.com/kubernetes/kubernetes/blob/v1.7.11/test/images/nvidia-cuda/Dockerfile
      image: "k8s.gcr.io/cuda-vector-add:v0.1"
      resources:
        limits:
          nvidia.com/gpu: 1 # requesting 1 GPU
```

Note the `resources.limits`.  This field has very particular restrictions that are different from CPUs. These are listed as 

> * GPUs are only supposed to be specified in the limits section, which means:
> * You can specify GPU limits without specifying requests because Kubernetes will use the limit as the request value by default.
> * You can specify GPU in both limits and requests but these two values must be equal.
> * You cannot specify GPU requests without specifying limits.
> * Containers (and pods) do not share GPUs. There’s no overcommitting of GPUs.
> * Each container can request one or more GPUs. It is not possible to request a fraction of a GPU.

I think most of these changes will be made in the request struct and in the stack file schema https://github.com/openfaas/faas/blob/master/gateway/requests/requests.go#L47 and https://github.com/openfaas/faas-cli/blob/master/stack/schema.go#L50

Modifying the `FunctionRequests` struct would be the absolute minimum required change.

To support the mixed GPU case, we need to support allowing the developer to specify a `nodeSelector`, e.g.

```
apiVersion: v1
kind: Pod
metadata:
  name: cuda-vector-add
spec:
  restartPolicy: OnFailure
  containers:
    - name: cuda-vector-add
      image: "k8s.gcr.io/cuda-vector-add:v0.1"
      resources:
        limits:
          nvidia.com/gpu: 1
  nodeSelector:
    accelerator: nvidia-tesla-p100
```

This would be adding a new option to the http requests and the stack schema.

**Comment by stefanprodan** [comment]
We already cover the node selector via stack constraints.

PS. I think this issue should be moved to faas-netes since it's Kubernetes specific.

**Comment by alexellis** [comment]
I would like project research and initiatives to start out here in the FaaS repo for visibility. 

Thanks for the comments Lucas.

**Comment by dkozlov** [comment]
FYI: https://github.com/dkozlov/openfaas-tensorflow-gpu

**Comment by alexellis** [comment]
That project looks like a useful example. 

* I can’t see a patch for the GPU. Does it “just work”? 
* where is the source for the object detection function?
* is it demonstrably faster on GPU vs. CPU?

**Comment by alexellis** [comment]
@dkozlov we had some discussion about this on Slack.. please can you summarize the points for the community?

**Comment by alexellis** [comment]
> > could you run two different functions using GPU at the same time?
> Yes, I can

I'm confused by this comment - I thought we were talking about scheduling constraints on Slack because two Pods cannot use the same GPU at the same time?

**Comment by dkozlov** [comment]
Also I have found following problems with native [Schedule GPUs](https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/):

> Containers (and pods) do not share GPUs. There’s no overcommitting of GPUs. - Each container can request one or more GPUs. It is not possible to request a fraction of a GPU.

As workaround I have implemented following:
* Add label to GPU nodes (```sudo kubectl label nodes  ip-123-45-67-890 label=gpu```) and to your OpenFaaS function
* Specify --max-pods limit for your node [https://github.com/kubernetes/kubernetes/issues/23349](https://github.com/kubernetes/kubernetes/issues/23349)

**Comment by dkozlov** [comment]
```I'm confused by this comment - I thought we were talking about scheduling constraints on Slack because two Pods cannot use the same GPU at the same time?```

If you install only NVIDIA drivers, docker and nvidia-docker it is enough to start GPU docker containers by kubernetes without any device plugin.

**Comment by alexellis** [comment]
My question was: "could you run two different functions using [the same] GPU at the same time?" (expecting an answer of no) and you answered "Yes, I can". Are we talking about the same thing? I thought GPUs could only be used by a single container/Pod at a time?

**Comment by dkozlov** [comment]
I could repeat it again: "Yes, it possible" :). It was even possible in 2016, see [ClarifAI blog post](https://clarifai.com/blog/scale-your-gpu-cloud-infrastructure-with-kubernetes) ```allow multiple pods on the same machine to share the same card, even if you know what you’re doing (at least on paper: ask us about this one weird trick to do just that!)```. Have you checked https://github.com/dkozlov/openfaas-tensorflow-gpu manual?

**Comment by dkozlov** [comment]
```My question was: "could you run two different functions using [the same] GPU at the same time?" (expecting an answer of no) and you answered "Yes, I can". Are we talking about the same thing? I thought GPUs could only be used by a single container/Pod at a time?```
If you try [nvidia-docker](https://github.com/NVIDIA/nvidia-docker/) you could allow single GPU usage by more than one container at a time.

**Comment by dkozlov** [comment]
> This is correct in the scope of Kubernetes, GPU resources are integer values and will belong to a single container. Unless you try to hack around it, that is :) In the Kubernetes issue linked above, I was trying to pitch the idea of sharing a GPU across all the containers in a single pod.
@flx42 Which another tricks/hacks we could do for overcommitting of GPUs in scope of Kubernetes? I am asking because OpenFaaS scales by pods and it could not be scaled by containers in a single pod.

**Comment by flx42** [comment]
I don't think you should try to hack around the official upstream support: that means don't overcommit GPUs. 

If you need to run multiple pods for the same function, you will need multiple GPUs.

**Comment by dkozlov** [comment]
FYI: https://github.com/Microsoft/KubeGPU seems that Microsoft trying to solve this problem

**Comment by alexellis** [comment]
@flx42 thanks for your input :+1: I would like to figure out what we need to do in the project to make it easy to consume GPU in a function on GKE or a bare-metal node / VM with nvidia-docker swapped in. If you'd like to collaborate on this we are also talking on Slack.

**Comment by flx42** [comment]
I think you should embrace the current upstream support, including its limitations.
If you assume that the cluster is already configured with the NVIDIA device driver, the device plugin and optionally taints/tolerations (see [this article](https://notes.mindprince.in/2017/12/17/dedicated-node-pools-and-ExtendedResourceToleration-admission-controller.html)), then you can just schedule pods consuming resources of type `nvidia.com/gpu`.

For the sake of simplicity and to avoid falling into suboptimal scheduling corner cases, I think you should limit the initial implementation to 1 GPU per container. i.e. `nvidia.com/gpu: 1`

**Comment by alexellis** [comment]
Would either of you be interested in helping to implement that within the project?

**Comment by feri** [comment]
I'm working on a patch that will enable scheduling functions in k8s if there is an extended resource exposed, by let's say a suitable device plugin, such as [[this](https://github.com/intel/intel-device-plugins-for-kubernetes)]. The work includes changes to faas-netes and faas-cli, and a minor one to the[ FunctionResources struct](https://github.com/openfaas/faas/blob/08d57de1f6950a5700f0443f32d42f0e34ecb35b/gateway/requests/requests.go#L47) that is from faas. Naturally faas-cli and faas patches will not be k8s specific.

**Comment by dkozlov** [comment]
> Do you or @dkozlov have a good container or some sample code that can verify that it has used or is using a GPU? 

@alexellis https://hub.docker.com/r/tensorflow/tensorflow/
 nvidia-docker run -it -p 8888:8888 tensorflow/tensorflow:latest-gpu
 or
sample code:
```
python3 -c "from tensorflow.python.client import device_lib; print(device_lib.list_local_devices())"
```

**Comment by sberryman** [comment]
Will this work on hosts with >1 GPU? I have a computer with two GTX 1080 TI's that I use for training or bulk inference. NVIDIA allows you to peg a docker container to a single GPU via an environment variable. `NVIDIA_VISIBLE_DEVICES=0` would restrict that container to the first GPU while `NVIDIA_VISIBLE_DEVICES=1` goes to GPU with index 1, etc.

**Comment by flx42** [comment]
@sberryman yes, our device plugin implementation supports multiple GPUs on one node and set this environment variable accordingly for the container.

**Comment by zouyee** [comment]
https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v1.11/nvidia-device-plugin.yml

**Comment by alexellis** [comment]
Bump

**Comment by vielmetti** [comment]
Bringing this to @dieterreuter attention, with the Jetson Nano as a target device for experimentation.

**Comment by alexellis** [comment]
@johnmccabe rebuilt his kennel to use the GPU in Docker.

**Comment by vielmetti** [comment]
`docker run --device=/dev/nvhost-ctrl --device=/dev/nvhost-ctrl-gpu --device=/dev/nvhost-prof-gpu --device=/dev/nvmap --device=/dev/nvhost-gpu --device=/dev/nvhost-as-gpu -v /usr/lib/aarch64-linux-gnu/tegra:/usr/lib/aarch64-linux-gnu/tegra device_query` - successfully run per instructions from https://github.com/Technica-Corporation/Tegra-Docker

**Comment by aimbot31** [comment]
Hello, is it still a subject going ? Thank's for the answers

**Comment by LucasRoesler** [comment]
@aimbot31 I think the best wait to achieve this is through the new Profiles feature in faas-netes https://docs.openfaas.com/reference/profiles/#use-tolerations-and-affinity-to-separate-workloads

Using profiles exposes the ability to give node affinity to your functions so that they run on nodes with a GPU available

**Comment by guptaprakash9** [comment]
@LucasRoesler I am not able to get Profiles method to work on GKE. It allows me to run my pod on GPU node fine, but apparently GKE doesn't attach and make GPU available to a pod unless explicitly requested via limits. Since its a managed Kubernetes, I cannot configure default runtime to nvidia-docker on these nodes. Any suggestions?

**Comment by LucasRoesler** [comment]
@guptaprakash9 unfortunately, i don't have a good answer. I can't think of a good way to set that request in OpenFaaS right now

@alexellis we either need to expand the spec for requests _or_ add a custom profiler field to configure this 

the relevant GKE docs https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#pods_gpus

**Comment by alexellis** [comment]
@guptaprakash9 if you or anyone landing here still has interest, we would be open to having this feature a) sponsored b) contributed or just c) [written up in detail as a proposal](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#i-have-a-great-idea)

https://github.com/openfaas/faas/issues/639#issuecomment-742263618

**Comment by rajitha1998** [comment]
Hello everyone, I would like to implement a component for OpenFaaS which will help to accelerate functions with GPU and TPU.  This will benefit in scientific applications, video frame analysis tasks etc.  Any ideas/suggestions on what kind of a component will be best fitting to this project?

**Comment by rajitha1998** [comment]
Found something interesting @alexellis. 
* https://github.com/AliyunContainerService/gpushare-scheduler-extender

*  [More info](https://ucc.alicdn.com/download/%E9%A6%96%E4%B8%AA%E6%99%AE%E6%83%A0%E7%A4%BE%E5%8C%BA%E7%9A%84%E5%B9%B3%E6%B0%91%E5%8C%96%E6%96%B9%E6%A1%88-GPU%E5%85%B1%E4%BA%AB%E8%B0%83%E5%BA%A6.pdf)

**Comment by alexellis** [comment]
Thanks for the link @rajitha1998 

You could try starting with [faasd](https://github.com/openfaas/faasd) which is fairly small and hackable, you can configure containerd to use a local GPU or TPU that you have. It's on my list and it's not for lack of interest, but since we are way [off the funding target on GitHub Sponsors](https://github.com/sponsors/openfaas/), and I have no access to such equipment, it's up to the community to invest their own time and resources into this.

**Comment by rajitha1998** [comment]
I am currently a bit busy with my university work. But I will add anything here which helps. Will look into this more when I get time :) @alexellis. 

Here there is a way to create a local Kubernetes cluster with GPU with MiniKube but a Linux computer is required.

Using the ‘none’ driver seems to be the easy way: https://minikube.sigs.k8s.io/docs/tutorials/nvidia_gpu/

Having a low-level  GPU such as Nvidia 940MX is enough: https://developer.nvidia.com/cuda-gpus

**Comment by bbean86** [comment]
I've been able to get a function working with a GPU attached on GKE, but I have to manually edit the Deployment and add the limit they expect for `nvidia.com/gpu`. Since this limit isn't one of cpu or memory I can't add it through stack.yml, and it gets reset every time I deploy the function. It would be nice to be able to submit this limit and have it work without needing to edit the deployment. 
Here's the link to the GKE docs showing an example of this limit: https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#pods_gpus 

I also had to create a custom template with the CUDA libraries installed and make sure the GKE cluster had the Nvidia driver installer daemon deployed. Not sure how much of this work is useful to the community but I'd be happy to share what's useful.
